### PR TITLE
New mfa rules

### PR DESCRIPTION
**Pull request by topherMarie**

### CODE REVIEW

File: rules.json [hunk 1]
```diff
@@ -482,6 +500,24 @@
...
+      },
+      {
+        "title": "Force multifactor when country has changed",
+        "categories": [
+          "multifactor",
+          "guardian"
+        ],
+        "description": "<p>This rule is used to trigger multifactor authentication when the user has changed countries since the last request..</p>\n",
+        "code": "function (user, context, callback) {\n\n  user.app_metadata = user.app_metadata || {}; \n  \n  if(user.app_metadata.last_location !== context.request.geo  ip.country_code){\n    context.multifactor = { \n      provider: 'guardian',\n      ignoreCookie: true, \n    };  \n  }\n  \n  //Set the location context for next time\n    user.app_metadata.last_location = context.request.geoip.country_code;\n  auth0.users.updateAppMetadata(user.user_id, user.app_metadata)\n    .then(function(){\n      cal  lback(null, user, context);\n    })  \n    .catch(function(err){\n      callback(err);\n  }); \n}\n"
```
Comment (topherMarie) [inline rules.json]: Here is a copy paste into vi, followed by :1,$s/\\n/^M/g

function (user, context, callback) {

  user.app_metadata = user.app_metadata || {}; 
  
  if(user.app_metadata.last_location !== context.request.geo  ip.country_code){
    context.multifactor = { 
      provider: 'guardian',
      ignoreCookie: true, 
    };  
  }
  
  //Set the location context for next time
    user.app_metadata.last_location = context.request.geoip.country_code;
  auth0.users.updateAppMetadata(user.user_id, user.app_metadata)
    .then(function(){
      cal  lback(null, user, context);
    })  
    .catch(function(err){
      callback(err);
  }); 
}

File: rules.json [hunk 2]
```diff
@@ -459,6 +459,24 @@
...
       },
       {
+        "title": "Force multifactor when IP outside an expected range",
+        "categories": [
+          "multifactor",
+          "guardian"
+        ],
+        "description": "<p>This rule is used to trigger multifactor authentication when the requesting IP is from outside the corporate IP range.</p>\n",
+        "code": "function (user, context, callback) {\n  var ipaddr = require('ipaddr.js');\n  var corp_network = '192.168.1.134/26';\n\n  var current_ip = ipaddr.parse(c  ontext.request.ip); \n  if(!current_ip.match(ipaddr.parseCIDR(corp_network))){\n    context.multifactor = {\n        provider: 'guardian',\n        ignoreCookie: true, \n        };  \n  }\n  \n  callback(null, user, context);\n}\n"
```
Comment (cristiandouce) [inline rules.json]: @tjanczuk @topherMarie fyi, the `rules.json` is autogenerated from this script: https://github.com/auth0/rules/blob/master/build

and it runs with every `npm version {version}` command as shown here: https://github.com/auth0/rules/blob/master/package.json#L7

So there is no need to submit the `rules.json` file with every commit.

File: rules/guardian-multifactor-ip-range.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+function (user, context, callback) {
+  var ipaddr = require('ipaddr.js');
+  var corp_network = "192.168.1.134/26";
+
+  var current_ip = ipaddr.parse(context.request.ip); 
+  if(!current_ip.match(ipaddr.parseCIDR(corp_network))){
+    context.multifactor = {
+        provider: 'guardian',
+        ignoreCookie: true, 
+      };
```
Comment (cristiandouce) [inline rules/guardian-multifactor-ip-range.md]: here indentation seems odd

File: rules/guardian-multifactor-ip-range.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+```js
+function (user, context, callback) {
+  var ipaddr = require('ipaddr.js');
+  var corp_network = "192.168.1.134/26";
+
+  var current_ip = ipaddr.parse(context.request.ip); 
+  if(!current_ip.match(ipaddr.parseCIDR(corp_network))){
+    context.multifactor = {
+        provider: 'guardian',
+        ignoreCookie: true,
```
Comment (cristiandouce) [inline rules/guardian-multifactor-ip-range.md]: I'd remove the extra `,` in the end, as per our `javascript` conventions: https://github.com/auth0/javascript

File: rules/guardian-multifactor-changed-location.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  
+  //Set the location context for next time
+  user.app_metadata.last_location = context.request.geoip.country_code;
+  auth0.users.updateAppMetadata(user.user_id, user.app_metadata)
+    .then(function(){
+      callback(null, user, context);
+    })  
+    .catch(function(err){
+      callback(err);
+  });
```
Comment (cristiandouce) [inline rules/guardian-multifactor-changed-location.md]: odd indentation here.

File: rules/guardian-multifactor-changed-location.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+```js
+function (user, context, callback) {
+
+  user.app_metadata = user.app_metadata || {}; 
+  
+  if(user.app_metadata.last_location !== context.request.geoip.country_code){
+    context.multifactor = { 
+      provider: 'guardian',
+      ignoreCookie: true,
```
Comment (cristiandouce) [inline rules/guardian-multifactor-changed-location.md]: remove extra `,` at the end

File: rules/guardian-multifactor-changed-location.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+## Multifactor when country changes
+
+This rule is used to trigger multifactor authentication when the current country does not match the country of the previous session
+
+```js
+function (user, context, callback) {
+
+  user.app_metadata = user.app_metadata || {}; 
+  
+  if(user.app_metadata.last_location !== context.request.geoip.country_code){
```
Comment (cristiandouce) [inline rules/guardian-multifactor-changed-location.md]:
```js
if (check) {
```
> check the spaces ☝️

File: rules/guardian-multifactor-changed-location.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    context.multifactor = { 
+      provider: 'guardian',
+      ignoreCookie: true, 
+    };  
+  }
+  
+  //Set the location context for next time
+  user.app_metadata.last_location = context.request.geoip.country_code;
+  auth0.users.updateAppMetadata(user.user_id, user.app_metadata)
+    .then(function(){
```
Comment (cristiandouce) [inline rules/guardian-multifactor-changed-location.md]:
```js
function () {
```
check the spaces ☝️

File: rules.json (hunk 2 above)
Comment (topherMarie) [inline rules.json]: I did not know that, very useful information, thanks!

**Pull request merged**